lib/elixir/lib/kernel.ex

```diff
@@ -2872,6 +2872,8 @@ defmodule Kernel do
         :lists.foldr(fn x, acc ->
           quote do: :erlang.or(unquote(comp(left, x)), unquote(acc))
         end, comp(left, h), t)
+      {:++, _meta, [h, t]} ->
+        quote do: :erlang.or(unquote(left) in unquote(t), unquote(left) in unquote(h))
```

Comment:
The difference will be in that currently `1 in [1] ++ nil` won't raise any error and with patch it will raise.
Anyway, I think we shouldn't rewrite `++` case as it is illegal expression in guards on it's own.
Also I believe this patch will make the error message for `123 ++ []` case worse as well.